app/Http/Controllers/TaskController.php

```diff
@@ -0,0 +1,165 @@
+            $tasks = Task::where('sender_user_id', $user->id)->get()->sortBy('created_at');
+        } elseif ($activeFilter == 'archived') {
+            $tasks = Task::where('recipient_user_id', $user->id)->whereIn('status', [TaskStatus::COMPLETED->value, TaskStatus::DECLINED->value])->get()->sortBy('created_at');
+        } else {
+            $tasks = Task::where('recipient_user_id', $user->id)->where('status', TaskStatus::PENDING->value)->get()->sortBy('created_at');
+        }
+
+        return view('tasks.index', compact('tasks', 'activeFilter'));
+    }
+
+    /**
+     * Store a newly created task in storage.
+     */
+    public function store(Request $request)
+    {
+
+        $this->authorize('create', Task::class);
+
+        $data = $request->validate([
+            'type' => ['required', new ValidTaskType],
+            'message' => 'sometimes|min:3|max:256',
+            'reference_user_id' => 'required|exists:users,id',
+            'reference_training_id' => 'required|exists:trainings,id',
+            'recipient_user_id' => 'required|exists:users,id',
+        ]);
+
+        $data['sender_user_id'] = auth()->user()->id;
+        $data['created_at'] = now();
+
+        // Check if recipient is mentor or above
+        $recipient = User::findOrFail($data['recipient_user_id']);
+
+        // Policy check if recpient can recieve a task
+        if ($recipient->can('receive', Task::class)) {
+            // Create the model
+            $task = Task::create($data);
+
+            // Run the create method on the task type to trigger type specific actions on creation
+            $task->type()->create($task);
+
+            return redirect()->back()->with('success', 'Task created successfully.');
+        }
+
+        return redirect()->back()->withErrors('Recipient is not allowed to receive tasks.');
+
+    }
+
+    /**
+     * Complete the specified task
+     *
+     * @param  \App\Models\Task  $task
+     */
+    public function complete(Request $request, int $task)
+    {
+
+        $this->authorize('update', Task::class);
+
+        $task = Task::findOrFail($task);
+
+        $task->status = TaskStatus::COMPLETED->value;
+        $task->closed_at = now();
+        $task->save();
+
+        // Run the complete method on the task type to trigger type specific actions on completion
+        $task->type()->complete($task);
+
+        return redirect()->back();
+    }
+
+    /**
+     * Decline the specified task
+     *
+     * @param  \App\Models\Task  $task
+     */
+    public function decline(Request $request, int $task)
+    {
+
+        $this->authorize('update', Task::class);
+
+        $task = Task::findOrFail($task);
+
+        $task->status = TaskStatus::DECLINED->value;
+        $task->closed_at = now();
+        $task->save();
+
+        // Run the decline method on the task type to trigger type specific actions on decline
+        $task->type()->decline($task);
+
+        return redirect()->back();
+    }
```

Comment:
suggestion: contract the stuff a little bit and re-use a smidge. Arguably, the close functionality should be moved *into* the task model itself.

```suggestion
    /**
     * Close the specified task with a given status.
     */
    protected function close(Task|int $task, TaskStatus $status): Task
    {
        $this->authorize('update', Task::class);
        $task = Task::findOrFail($task);
        $task->status = $status;
        $task->closed_at = now();
        $task->save();

        return $task;
    }

    /**
     * Complete the specified task.
     */
    public function complete(Request $request, Task|int $task): RedirectResponse
    {
        $completed = self::close($task, TaskStatus::COMPLETED);

        // Run the complete method on the task type to trigger type specific actions on completion
        $completed->type()->complete($completed);

        return redirect()->back()->with('success', sprintf('Completed task regarding %s from %s.', $completed->subject->name, $completed->creator->name));
    }

    /**
     * Decline the specified task.
     */
    public function decline(Request $request, Task|int $task): RedirectResponse
    {
        $declined = self::close($task, TaskStatus::DECLINED);

        // Run the decline method on the task type to trigger type specific actions on decline
        $declined->type()->decline($declined);

        return redirect()->back()->with('success', sprintf('Declined task regarding %s from %s.', $declined->subject->name, $declined->creator->name));
    }

```

Reply:
I'll keep it in the controller for now, as we're usually doing this kind of model manipulation in the controller in the rest of the application. So that's rather a own 🐰 🕳️ to map out and clean up overall.